### ISSUE TITLE
Server Classes exposed to web applications - Jetty 9.2.22.v20170606

### ISSUE THREAD

**Issue by gjafachini** [eclipse/jetty.project#1925]
We have recently upgrade our Jetty cluster from 8 to 9.2.22 and we deploy several webapps into different containers (at least 12 containers in production).

The modules that we have enable with default configuration at start.ini for all containers are:
deploy
websocket
ext (nothing at lib/ext/)
jsp
apache
http
jmx
servlets
logging

We have only 2 webapps that uses websockets, one due to this dep-tree.

The **first one** is working fine and is used to provide a task status service through websockets directly to web browser clients.
It depends on:
```
[INFO] leApp1-app:jar:4.10.00-SNAPSHOT
[INFO] +- org.cometd.java:cometd-java-annotations:jar:3.0.10:compile
[INFO] |  +- javax.inject:javax.inject:jar:1:compile
[INFO] |  +- javax.annotation:jsr250-api:jar:1.0:compile
[INFO] |  +- org.cometd.java:cometd-java-server:jar:3.0.10:compile
[INFO] |  |  +- org.cometd.java:cometd-java-common:jar:3.0.10:compile
[INFO] |  |  |  \- org.eclipse.jetty:jetty-util-ajax:jar:9.2.18.v20160721:compile
[INFO] |  |  \- org.eclipse.jetty:jetty-jmx:jar:9.2.22.v20170606:compile
[INFO] |  +- org.eclipse.jetty:jetty-util:jar:9.2.18.v20160721:compile
[INFO] |  \- org.slf4j:slf4j-api:jar:1.6.6:compile
```
and
```
<dependency>
	<groupId>org.cometd.java</groupId>
	<artifactId>cometd-java-websocket-javax-server</artifactId>
	<version>3.0.10</version>
</dependency>
```

The second one is breaking when it tries to create an SslContextFactory with this stack trace:

```
java.lang.LinkageError: loader constraint violation: 
when resolving method "org.eclipse.jetty.websocket.client.WebSocketClient.<init>(Lorg/eclipse/jetty/util/ssl/SslContextFactory;)V"
the class loader (instance of org/eclipse/jetty/webapp/WebAppClassLoader) of the current class, com/.../MonitoringPlatformImpl, and 
the class loader (instance of org/eclipse/jetty/start/Classpath$Loader) for the method's defining class,
org/eclipse/jetty/websocket/client/WebSocketClient, have different Class objects for the type org/eclipse/jetty/util/ssl/SslContextFactory used in the signature
```

It depends on:
```
<dependency>
	<groupId>org.eclipse.jetty.websocket</groupId>
	<artifactId>websocket-api</artifactId>
</dependency>
<dependency>
	<groupId>org.eclipse.jetty</groupId>
	<artifactId>jetty-util</artifactId>
</dependency>
<dependency>
	<groupId>org.eclipse.jetty.websocket</groupId>
	<artifactId>javax-websocket-client-impl</artifactId>
</dependency>
```

The first thing we tried, was to add scope provided to the dependencies, however, we began to have this failure:
```
java.lang.NoClassDefFoundError: org/eclipse/jetty/util/ssl/SslContextFactory
...
Caused by:
java.lang.ClassNotFoundException: org.eclipse.jetty.util.ssl.SslContextFactor
...
```

And then, we tried to disable websocket module and provide all the websockets dependencies to the first module, but it began to fail, missing, annotation, and then, enabling annotation, it missed jetty-io, and we kept enabling stuff but wasn't enough.

We found this email here with the exact same problem: [http://dev.eclipse.org/mhonarc/lists/jetty-dev/msg03096.html]

And we tried a number of different WebAppContext configurations at WEB-INF/jetty-web.xml, with addServerClass, prependServerClass, enabling/disabling classpaths but nothing solved.

From our understatement of the docs, the server module classloader shouldn't expose anything, consequentially, we shouldn't been seeing the linkage error on jetty-util package in the first place. If jetty-util.jar is only available at server level, how are we conflicting both packages, from jetty server and from the webapp/lib ? Are we missing something here?

At the email to the jetty-dev group Christian Elsen attached a simple example, and it fits perfectly the issue that we are facing. We don't want to change the default contexts of jetty, since there are lots of webapps relying on that previous setup, and we shouldn't change the base container configuration because of one webapp dependency. Btw, with the same modules enabled, at Jetty 8, we didn't had this problem.

Thx

**Comment by gregw** [comment]
Firstly 9.2.22 is an old release that is past end-of-life.  I assume you are using it because you are using java7, which is also end-of-life and thus if you are using SSL, it is not secure because exploits have been discovered in SSL since.

It is true that technically no server classes should be exposed, however practically we must leave a few holes in the classloader for various features whose spec (such as websocket) didn't deal with the classloading issues of app servers.

We have adjusted the default classloader configuration over the 9.2,9.3 and 9.4 releases, so there is some variation of behaviour.  In 10, we are removing the static classpath configuration and each module will condition the classpath as it requires it to be... hopefully resulting in less inappropriate holes.

Also we are now more often using a shaded or hybrid shaded version of the jetty client, so it does not clash with any server classes that may be exposed.  But the shaded version is not available in 9.2.22 by default, so if you want to go this path you will need to shade the dependencies yourself (easy enough with the maven shade plugin).

Alternately, you can manually adjust the system and server classes to achieve the class hiding you need.  Do a `server.dumpStdErr()` after startup (there is an option in server.ini or jetty.ini for this) and you should be able to see the system and server classes.   You can then adjust those settings so that the server classes are correctly hidden.

**Comment by gjafachini** [comment]
Hi Greg,
thanks very much for your response, indeed we have migrated to 9.2.22 to step closer to Java 8 and than migrate the java too.

Your explanation give us a good background to workaround or upgrade jetty version.
